test/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -12,6 +12,133 @@ function FiniteDifferences.to_vec(x::Val)
 end
 
 @testset "Factorizations" begin
+    @testset "lu decomposition" begin
+        # avoid implementing to_vec(::LU)
+        asnt(F::LU) = (U=F.U, L=F.L, factors=F.factors)
+        asnt2(F::LU) = (U=F.U, L=F.L)
+        n = 10
+        @testset "lu! frule" begin
+            @testset "lu!(A::Matrix{$T}, $pivot) for size(A)=($m, $n)" for
+                T in (Float64, ComplexF64),
+                pivot in (Val(true), Val(false)),
+                m in (7, 10, 13)
+
+                A = randn(T, m, n)
+                ΔA = rand_tangent(A)
+                F = lu!(copy(A), pivot)
+                ∂F_fd = jvp(_fdm, A -> asnt(lu!(copy(A), pivot)), (A, ΔA))
+                F_ad, ∂F_ad = @inferred frule(
+                    (Zero(), copy(ΔA), DoesNotExist()), lu!, copy(A), pivot
+                )
+                @test F_ad == F
+                @test ∂F_ad isa Composite{typeof(F)}
+                @test ∂F_ad.L ≈ ∂F_fd.L
+                @test ∂F_ad.U ≈ ∂F_fd.U
+                @test ∂F_ad.factors ≈ ∂F_fd.factors
```

Comment:
what is preventing us using ChainRulesTestUtils here?
Can we open an issue and cross-reference it in the code?
(or cross reverence existing one)

I am trying to workout where the limitations in ChainRulesTestUtils are so we can fix them

Reply:
It was the weirdness of us using the `L` and `U` fields of `Composite{LU}` instead of `factors`, when `L` and `U` are not fields of the primal. But I have now changed this and we can use ChainRulesTestUtils.